### ISSUE TITLE
Official support for Unity 2021 LTS

### ISSUE THREAD

**Issue by mrwellmann** [Azure/azure-spatial-anchors-samples#354]
I've seen in some issues ([1](https://github.com/Azure/azure-spatial-anchors-samples/issues/335#issuecomment-1107869579), [2](https://github.com/Azure/azure-spatial-anchors-samples/issues/339), [3](https://github.com/Azure/azure-spatial-anchors-samples/issues/339)) people saying they are using or testing Unity 2021.
- Are there any plans to officially support Unity 2021. This would also mean to update AR Foundation and all other packages mentioned in the [Dokumentation ](https://docs.microsoft.com/de-de/azure/spatial-anchors/how-tos/setup-unity-project?tabs=xr-plugin-framework%2Cunity-2020%2Cunity-package-web-ui%2CExtraConfigurationsHoloLens)?
- To other unity developers have you already tried ASA with Unity 2021?

What I can say from my point is that we are using Unity 2020 and AR Foundation 4.2.0 including ARCore / ARKit 4.2.0 since a while now without issues.

**Comment by SimonDarksideJ** [comment]
Any comment @darax ?  or suggestions about getting this sample to run in Unity 2021?

Especially given Unity have REMOVED the WIndowsMR classes from Unity 2021 that the sample uses, likely due to the migration to OpenXR.

Ideally, this sample should still be able to run without MRTK3, given the Android / iOS support.

**Comment by darax** [comment]
@mrwellmann, I missed seeing this earlier.
@SimonDarksideJ  Our next release (we're finalizing some things, but looks like Mid July) will work with the Unity 2021.3 LTS, though you will need to migrate from Windows MR to OpenXR if you are still on Windows MR.

**Comment by SimonDarksideJ** [comment]
Thanks for the feedback @darax
I've already updated and upgraded the project to 2021
Although I could do with some tips on getting hands select actions working for Hololens 2 in openxr, as I can't seem to get the airtap style working. (and I wouldn't want to add the entire mrtk3 just for a button) 
Was about to resort to a good old hl1 gaze activation in desperation.

**Comment by darax** [comment]
@SimonDarksideJ  
What worked for me was taking WindowsMR Gestures script and reimplementing it for openxr.

```
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.XR;
using System;

namespace UnityEngine.XR.WindowsMR
{
    public class WindowsMRGestures : MonoBehaviour
    {
        public event Action<WindowsMRTappedGestureEvent> onTappedChanged;
        bool[] wasClicked = new bool[2] { false, false };

        void Update()
        {
            bool rightClicked = HandClicked(XRNode.RightHand);
            bool leftClicked = HandClicked(XRNode.LeftHand);

            if(rightClicked && rightClicked != wasClicked[0] || leftClicked && leftClicked != wasClicked[1])
            {
                onTappedChanged?.Invoke(new WindowsMRTappedGestureEvent());
            }

            wasClicked[0] = rightClicked;
            wasClicked[1] = leftClicked;
        }

        private bool TryGetIsGrabbing(InputDevice device)
        {
            bool isGrabbing = false;

            if (device.TryGetFeatureValue(CommonUsages.triggerButton, out isGrabbing))
            {
                return isGrabbing;
            }
            else if (device.TryGetFeatureValue(CommonUsages.gripButton, out isGrabbing))
            {
                return isGrabbing;
            }
            else if (device.TryGetFeatureValue(CommonUsages.primaryButton, out isGrabbing))
            {
                return isGrabbing;
            }
            return isGrabbing;
        }

        bool HandClicked(XRNode handNode)
        {
            bool Clicking = false;
            InputDevice handDevice = InputDevices.GetDeviceAtXRNode(handNode);
            if (handDevice != null)
            {
                bool deviceValid = handDevice.TryGetFeatureValue(CommonUsages.isTracked, out bool deviceIsTracked);
                if (deviceValid && deviceIsTracked)
                {
                    Clicking = TryGetIsGrabbing(handDevice);
                }
            }

            return Clicking;

        }
    }

    public class WindowsMRTappedGestureEvent
    {

    }
}
```

**Comment by Varnauld** [comment]
Hi @mrwellmann @SimonDarksideJ - I think this was resolved with the last update this summer but want to make sure you don't have any further questions before we close this issue. Thanks!

**Comment by mrwellmann** [comment]
@Varnauld I think there is still some work to do here
- We are using Unity 2021 LTS with AR Foundation 4.2.3 for iOS and Android successfully. &#x2705;
- The [Dokumentation ](https://learn.microsoft.com/de-de/azure/spatial-anchors/how-tos/setup-unity-project?tabs=xr-plugin-framework%2Cunity-2020%2Cunity-package-web-ui%2CExtraConfigurationsHoloLens) is sill outdated.
- Can't say anything about the WindowsMR issues.

**Comment by Varnauld** [comment]
Thank you - tagging @darax for the documentation update

**Comment by sharramon** [comment]
Hey, the [documentation](https://learn.microsoft.com/en-us/azure/spatial-anchors/how-tos/setup-unity-project?tabs=xr-plugin-framework%2Cunity-2020%2Cunity-package-web-ui%2CExtraConfigurationsHoloLens) still has no mention of Unity 2021.3 LTS.

I'm using Unity 2021.3.12f1 and the Hololens2 (updated to most recent software at time of posting).

I have a few issues.

When I'm trying to run the sample scene after downloading the required packages, the sample scene still looks for WindowsMRGestures and throws an error. I've just patched in @darax 's solution posted above to work around that.

I also haven't been able to get spatial anchors to sync across sessions or devices, and the placed anchors appears seemingly at random. I've tried using extremely easy to notice colour coded points of interest to see if that was the problem but that hasn't seemed to work either.

How do you recommend I start debugging this?

**Comment by darax** [comment]
Hi @sharramon  We've addressed the unity 2021.3 issue in our next release.  We're doing some final fixes and preparing to test for a release around March 28th.

The recall issue might be related to the changes made to work around the windows MR deprecation.  I wonder if maybe the scene origin was accidentally moved? I'll paste some screenshots of the relevant bits, but do double check the project settings/xr settings are configured for OpenXR on hololens.
![image](https://user-images.githubusercontent.com/15933792/222315644-83320dd0-5d42-4e8b-ae50-baabf578d598.png)
![image](https://user-images.githubusercontent.com/15933792/222315708-61bea0bf-feb9-480f-ac38-6c8b2d7d1d21.png)

If that doesn't help, Unity writes a log file to the app's tempstate folder. You can get to it from the device portal website. I'd check that log and see if some errors are being logged that provide a clue.

**Comment by darax** [comment]
in 2.14.1
